### DISCUSSION TITLE
vs noodles and bedrs and rust-bio

### DISCUSSION THREAD

**Discussion by Kentalot** [quinlan-lab/bedder-rs#7] [General]
Hi,

I was looking into bed support in my rust code and saw that this repo, noodles, rust-bio, and bedrs had bed support.  I think the noodles and bedrs are aiming for bedtools-like abilities like this one.  Have you looked at the other repos and found any big differences?  What advantages and disadvantages do you think your repo has?  Thanks a lot and I'll keep an eye on the project.

**Comment by brentp** [comment]
Hi, 
noodles is quite different as it parses BED (and VCF and BAM/CRAM). bedder-rs uses noodles extensively. I think the same is true for rust-bio.

bedder-rs and bedrs are more similar, but as I understand it, bedrs uses in memory data structures and bedder uses streaming to do the intersections. I am very interested in bedrs and will follow as it develops.

Now, the disadvantage of both libs is that they lack functionality to operate on a range of formats. To now, bedder only supports bed and VCF (though BAM/CRAM are easily added). Instead, I am focusing on how everything will look for the final user, including python bindings to the lib.

**Reply by Kentalot** [reply]
Ah cool.  I will be binding to python as well, how do you plan on doing that?   For now, we will be using pyO3, but if you are already doing that, maybe we can piggyback off your work or contribute to that section of the code.  In terms of noodles, we tried using that for their vcf library but we found rust-htslib outperformed it by a lot, but maybe we didn't do a thorough enough of a test.   For now, we are going with rust-htslib, but it lacks bed support, which is why we were looking at different bed libraries.  Thanks!

**Reply by brentp** [reply]
yes, I have some bindings to noodles::vcf using pyo3. I didn't check performance, but indeed it's much harder to use, especially with pyo3 than rust-htslib. 
What project are you working on?

**Reply by Kentalot** [reply]
I'm working on converting [witty.er](https://github.com/Illumina/witty.er) to python.  The repo we're working on it is internal for now, but we've had some success with rust-htslib and vcf processing.  Just need some more bed support, but we're in the beginning of the process really.  We hope to finish by end of year though.

**Reply by brentp** [reply]
Ah, nice. I've used witty.er. 
Here is the stub of the noodles::vcf wrapper that I wrote to get help from the pyo3 community:
~~https://github.com/brentp/pyo3_ref/blob/pyref/src/lib.rs~~
https://github.com/brentp/pyo3_ref/blob/noodles/src/lib.rs

I'm hitting a lot of walls due to pyo3 not allowing lifetimes or generics. Those would indeed be avoided by using rust-htslib.

**Reply by brentp** [reply]
oops. wrong branch: https://github.com/brentp/pyo3_ref/blob/noodles/src/lib.rs

**Reply by brentp** [reply]
out of curiosity, did you benchmark noodles recently? I thought performance had improved lately.

**Reply by Kentalot** [reply]
I don't know how recent is recently, but we did some tests that ran a vcf (read and then modify and write) a few times through code using noodles and parallel code using rust-htslib maybe a month ago.  It's not a very thorough test but we just wanted to compare on a superficial level and we were leaning towards rust-htslib since htslib is well used and tested by the community.  So if it's similar or even slightly slower, we'd go with that.  Since our superficial results came out that rust-htslib was much faster, we decided to go with that in general.  We haven't did a test since though.  Maybe it's worth it to do it eventually, but for now, we will stick with one to finish the job first and we can always optimize later.

**Reply by brentp** [reply]
ok. that's recent. Maybe I'll re-consider as well.

**Reply by Kentalot** [reply]
you might want to do your own testing. I was looking over our code and the noodles code and htslib code is actually quite different so it is not necessarily exactly equivalent.  I don't want to spend the time to make them more equivalent, but I'll assume that even if we optimized the noodles code more, the performance is at least similar, which means rust-htslib for us still for now.

**Comment by Kentalot** [comment]
Actually after some more thought, I think we do require the support for intervals and intersecting and all that, which as you said, rust-bio doesn't provide.  Is bedder-rs ready to be used overall or it needs a lot more work before it's ready? The main README page seems to say that `downstream APIs to perform operations on the intersections` is not ready, but does that mean the intersections are ready to use?  If so, I'll try using this as a first pass.  Thanks!

**Reply by brentp** [reply]
I think it's ready. the main algo is very simple as it requires the inputs to be sorted, it's doing streaming intersections. 
You could even write your own impl of [Positioned](https://github.com/quinlan-lab/bedder-rs/blob/2a995f2138821b6a2532f4dde1cfd9f307a08c47/src/position.rs#L43-L52) for rust-htslib VCF 
and then write the [PositionedIterator](https://github.com/quinlan-lab/bedder-rs/blob/2a995f2138821b6a2532f4dde1cfd9f307a08c47/src/position.rs#L137-L155) to iterate over the file and yield your `Positioned` structs and be on your way. 

Here's everything you need to know about the Intersections:
https://github.com/quinlan-lab/bedder-rs/blob/2a995f2138821b6a2532f4dde1cfd9f307a08c47/src/intersection.rs#L38-L51

Those are public so you simply access them and match on the Position enum to get out your concrete type:
https://github.com/quinlan-lab/bedder-rs/blob/2a995f2138821b6a2532f4dde1cfd9f307a08c47/src/position.rs#L59-L68

In your case, you may need to use `Other(Box<dyn Positioned>)`, but to get feedback, I'd be willing to implement a *feature* that adds built-in support for rust-htslib VCF.

You'll hit some bumps, I'm sure, but I'll respond quickly. No worries if it's too much of a PITA, indeed it's early days for the library.
